src/mscorlib/src/System/String.cs

```diff
@@ -1208,6 +1272,42 @@ internal String[] SplitInternal(char[] separator, int count, StringSplitOptions
             }
             return foundCount;
         }
+
+        private struct SplitSeparatorArray
+        {
+            private readonly bool _single;
```

Comment:
The downside is it requires an extra check for `IsNullOrEmpty`, `Length`, and the indexer. It's not so bad for the first two, because they're only called once, but the indexer is called each time looping through the separators.

```c#
    private struct SplitSeparatorArray
    {
        private readonly string _separator;
        private readonly string[] _separators;

        public SplitSeparatorArray(string separator)
        {
            _separator = separator ?? string.Empty;
            _separators = null;
        }

        public SplitSeparatorArray(string[] separators)
        {
            _separator = null;
            _separators = separators;
        }

        public bool IsNullOrEmpty
        {
            get { return !IsSingle && (_separators == null || _separators.Length == 0); }
        }

        public int Length
        {
            get { return IsSingle ? 1 : _separators.Length; }
        }

        public string this[int index]
        {
            get { return IsSingle ? _separator : _separators[index]; }
        }

        private bool IsSingle
        {
            get { return _separator != null & _separator.Length != 0; }
        }
    }
```

FWIW, I originally had the following, which is more closely related to [`ParamsArray`](https://github.com/dotnet/coreclr/blob/cbf46fb0b6a0b209ed1caf4a680910b383e68cba/src/mscorlib/src/System/ParamsArray.cs). It keeps the struct down to 2 fields, but does require a one-time static (cached) `string[]` allocation, that I wanted to avoid. I felt the 3 fields vs 2 fields tradeoff was worth it to avoid the one-time allocation and minimize the amount of checks.

```c#
    private struct SplitSeparatorArray
    {
        // Sentinel fixed-length array eliminates the need for a "count" field keeping this
        // struct down to just 2 fields. This is only used for its "Length" property,
        // that is, its elements are never set or referenced.
        private static readonly string[] s_oneSeparatorArray = new string[1];

        private readonly string _separator0;

        // After construction, the first element of this array will never be accessed
        // because the indexer will retrieve this value from _separator0
        private readonly string[] _separators;

        public SplitSeparatorArray(string separator)
        {
            _separator0 = separator;

            // Always assign _separators to make use of its "Length" property
            _separators = s_oneSeparatorArray;
        }

        public SplitSeparatorArray(string[] separators)
        {
            _separator0 = separators != null && separators.Length > 0
                ? separators[0]
                : null;

            _separators = separators;
        }

        public bool IsNullOrEmpty
        {
            get { return _separators == null || _separators.Length == 0; }
        }

        public int Length
        {
            get { return _separators.Length; }
        }

        public string this[int index]
        {
            get { return index == 0 ? _separator0 : _separators[index]; }
        }
    }
```

I'll play around and see what works best and also consider the unsafe approach.